### PR TITLE
Extend inverses, and slightly clarify.

### PR DESCRIPTION
**Pull request by HansOlsson**
The main part of this change is to make it easier to specify inverses for functions that already invert other functions (see also the bugs https://github.com/modelica/ModelicaStandardLibrary/issues/3399 and https://github.com/modelica/ModelicaStandardLibrary/issues/3400 ).

The changes are:
- Allow arguments to be skipped in the inverse (reason provided).
- Remove the requirement that defaults should be the same. As we provide a complete inverse-call they are not used.
- Change the requirement that they should have the same formal parameters, to say that it is union of formal parameters and output that should remain the same.

Note: A missing feature is that for solve-functions it might be beneficial to say that the inverse function is "faster" in some way. Exactly how is not clear.

### CODE REVIEW

**Comment by thorade** [comment]
I support the idea to make it possible to inform the tool which one is the forward function and which one is a numerical inversion.

**Comment by HansOlsson** [comment]
An additional enhancement would be to not require that it is a function call.

E.g. https://github.com/modelica/ModelicaStandardLibrary/issues/3399 could be solved by `inverse(h=(if ... then h_T(T) else h_pT(p,T)))`

**Comment by HansOlsson** [comment]
> I support the idea to make it possible to inform the tool which one is the forward function and which one is a numerical inversion.

As I see there are two possibilities for this:

- Use another name than "inverse" for that: `annotation(otherInverse(h=h_pT(p,T))`
- Use a modifier on the output `annotation(inverse(h(someTag)=h_pT(p,T))`
Both would be ok, but require better names.

**Comment by HansOlsson** [comment]
> 
> 
> > I support the idea to make it possible to inform the tool which one is the forward function and which one is a numerical inversion.
> 
> As I see there are two possibilities for this:
> 
>     * Use another name than "inverse" for that: `annotation(otherInverse(h=h_pT(p,T))`
> 
>     * Use a modifier on the output `annotation(inverse(h(someTag)=h_pT(p,T))`
>       Both would be ok, but require better names.

A possibly better name would be "solve" (or solves?), as it solves that equation (numerically or otherwise).

**Comment by thorade** [comment]
Is the slow inverse always a numerical inverse? Would it be sufficient to use Inverse, AnalyticalInverse and NumericalInverse?

**Comment by HansOlsson** [comment]
> 
> 
> Is the slow inverse always a numerical inverse? Would it be sufficient to use Inverse, AnalyticalInverse and NumericalInverse?

I don't think that matters for the case I'm more considering, as it is specifying the inverse of the inverse, i.e., that the Analytic/Numeric inverse specifies the original function.
Specifying an analytic/numeric slow inverse might be possible, but it is a bit unclear when it is useful.

**Comment by HansOlsson** [comment]
Tangential issue:
Rename "formal parameter" to "input arguments" (except for one reference to normal CS) - to reduce confusion between parameter and parameters.
"The variables declared with input keyword are called input arguments in this document, and commonly referred to as formal parameters."
Agreement by acclamation.

Poll:
Extend inverse-annotation to this case:
Introduce different annotation, e.g. solve for this case:

**Comment by HansOlsson** [comment]
This should now be ready for review.

File: chapters/functions.tex [hunk 1]
```diff
@@ -1616,12 +1616,13 @@ \section{Declaring Inverses of Functions}\doublelabel{declaring-inverses-of-func
 parameters with and without default value. The restriction is that the
 \emph{number of unknown variables} (see \autoref{balanced-models}) in the output formal
 parameter of both $f_1$ and $f_2$ must be
-the same and that $f_2$ must have exactly the same formal
-parameters as $f_1$ (with the same defaults, if a formal
-parameter $u_m$ has a default), but the order of the formal
+the same and that $f_2$ should have an union of output and formal
```
Comment (henrikt-ma) [inline chapters/functions.tex:1619]: … should have **a** union

File: chapters/functions.tex [hunk 2]
```diff
@@ -1644,6 +1645,32 @@ \section{Declaring Inverses of Functions}\doublelabel{declaring-inverses-of-func
 \end{lstlisting}
 \end{example}
 
+The sub-set case is useful if $f_1$ computes the inverse of $f_2$ within a region, or up to a certain tolerance,
+the $f_1$ may specify $f_2$ as inverse with fewer arguments, skipping the arguments for tolerance and/or the region.
+
+{[}\emph{Example:}
```
Comment (henrikt-ma) [inline chapters/functions.tex:1651]: Use the `example` environment.

File: chapters/functions.tex [hunk 3]
```diff
@@ -1644,6 +1645,32 @@ \section{Declaring Inverses of Functions}\doublelabel{declaring-inverses-of-func
 \end{lstlisting}
 \end{example}
 
+The sub-set case is useful if $f_1$ computes the inverse of $f_2$ within a region, or up to a certain tolerance,
+the $f_1$ may specify $f_2$ as inverse with fewer arguments, skipping the arguments for tolerance and/or the region.
```
Comment (henrikt-ma) [inline chapters/functions.tex:1649]: Long sentence that is hard to parse.  What about this?
> The sub-set case is useful if $f_1$ computes the inverse of $f_2$ within a region, or up to a certain tolerance.  Then, $f_1$ may specify $f_2$ as inverse with fewer arguments, skipping the arguments for tolerance and/or the region.

File: chapters/functions.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/functions.tex]: Of course!
I even think I did at first, but then rebasing etc caused new conflicts.

File: chapters/functions.tex (hunk 3 above)
Comment (HansOlsson) [reply chapters/functions.tex]: Should be resolved now.
